src/libsystemd-network/sd-ndisc.c

```diff
@@ -524,24 +527,53 @@ static int ndisc_router_advertisment_recv(sd_event_source *s, int fd, uint32_t r
         else if (msg.msg_namelen != sizeof(sa.in6)) {
                 log_ndisc(nd, "Received invalid source address size from ICMPv6 socket: %zu bytes", (size_t)msg.msg_namelen);
                 return 0;
-        } else
+        }
+
+        if (sa.sa.sa_family == AF_INET6) {
+
+                if (sa.in6.sin6_scope_id) {
+                        ifindex = sa.in6.sin6_scope_id;
+
+                        if (ifindex != nd->index) {
+                                log_ndisc(nd, "Received RA on wrong interface: %d != %d. Ignoring.", ifindex, nd->index);
+                                return 0;
+                        }
+                }
+
                 gw = &sa.in6.sin6_addr;
+        }
 
         assert(!(msg.msg_flags & MSG_CTRUNC));
         assert(!(msg.msg_flags & MSG_TRUNC));
 
         CMSG_FOREACH(cmsg, &msg) {
-                if (cmsg->cmsg_level == SOL_IPV6 &&
-                    cmsg->cmsg_type == IPV6_HOPLIMIT &&
-                    cmsg->cmsg_len == CMSG_LEN(sizeof(int))) {
-                        int hops = *(int*)CMSG_DATA(cmsg);
 
-                        if (hops != 255) {
-                                log_ndisc(nd, "Received RA with invalid hop limit %d. Ignoring.", hops);
-                                return 0;
+                if (cmsg->cmsg_level == IPPROTO_IPV6) {
+
+                        switch (cmsg->cmsg_type) {
+
+                        case IPV6_PKTINFO: {
+                                struct in6_pktinfo *i = (struct in6_pktinfo*) CMSG_DATA(cmsg);
+
+                                ifindex = i->ipi6_ifindex;
+
+                                if (nd->index != ifindex) {
+                                        log_ndisc(nd, "Received RA on wrong interface: %d != %d. Ignoring.", ifindex, nd->index);
+                                        return 0;
+                                }
+                                break;
                         }
+                        case IPV6_HOPLIMIT: {
+                                int hops = *(int*) CMSG_DATA(cmsg);
```

Comment:
It wont' work as CMSG_DATA(cmsg) is a pointer.  We need to deference this value and store at hops.